test/types/instance.test-d.ts

```diff
@@ -6,6 +6,32 @@ import { HookHandlerDoneFunction } from '../../types/hooks'
 
 const server = fastify()
 
+server.decorate('nonexistent', () => {})
+server.decorateRequest('nonexistent', () => {})
+server.decorateReply('nonexistent', () => {})
+
+declare module '../../fastify' {
+  interface FastifyInstance {
+    functionWithTypeDefinition: (foo: string, bar: number) => Promise<boolean>
+  }
+  interface FastifyRequest {
+    numberWithTypeDefinition: number
+  }
+  interface FastifyReply {
+    stringWithTypeDefinition: 'foo' | 'bar'
+  }
+}
+expectError(server.decorate('functionWithTypeDefinition', (foo: any, bar: any) => {})) // error because invalid return type
+expectError(server.decorate('functionWithTypeDefinition', (foo: any, bar: any) => true)) // error because doesn't return a promise
+expectError(server.decorate('functionWithTypeDefinition', async (foo: any, bar: any, qwe: any) => true)) // error because too many args
+server.decorate('functionWithTypeDefinition', async (foo, bar) => true)
```

Comment:
Could you add a type assertion on the type of `foo` inside this function? Would be good to test that it is inferred properly